### ISSUE TITLE
no-unreachable didn't work with single if

### ISSUE THREAD

**Issue by jddxf** [eslint/eslint#8495]
**Tell us about your environment**

* **ESLint Version:** 3.19.0
* **Node Version:** 6.10.2
* **npm Version:**

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
const ERROR = 2

module.exports = {
    rules: {
      'no-unreachable': [ERROR]
    }
};
```

**What did you do? Please include the actual source code causing the issue.**

```js
function foo() {
  if (true) {
    return
  }
  return
}
```

**What did you expect to happen?**
ESLint report a no-unreachable error
**What actually happened? Please include the actual, raw output from ESLint.**
Lint passed

**Comment by alberto** [comment]
Did you expect an error because the test condition is always true?

**Comment by not-an-aardvark** [comment]
I'm not sure this is a bug; in general the code path analysis algorithm doesn't evaluate subexpressions to determine which code paths are reachable. With expressions like `true` it happens to be obvious, but it's not possible in general to determine whether any given expression will always evaluate to `true`.

**Comment by jddxf** [comment]
@alberto Yeah, but now it seems it shouldn't.

@not-an-aardvark Thanks for your explanation. I didn't think of that. Then is it a special case using loop constructs? For example, ESLint do report an error for following code block:
```
while (true) {
}
return
```

**Comment by not-an-aardvark** [comment]
Hmm, it looks like I was wrong -- we do evaluate literals such as `true` when doing code path analysis on loops (see [here](https://github.com/eslint/eslint/blob/d49acc309ad69d99db1e04e971c33a69fdf657da/lib/code-path-analysis/code-path-analyzer.js#L59-L75)). In that case I suppose it makes sense to use the same check for `if` statements as well.

**Comment by jddxf** [comment]
I'm not sure whether it makes sense. Seems that the code you refered to is used only to detect infinite loops. I can see that as ESLint don't report for 
```
while (false) {
  console.log()
}
return
```

**Comment by not-an-aardvark** [comment]
`while (false)`

![](https://i.imgur.com/SMEpnna.gif)

Hmm, this seems like something we should do, but if it's just used for detecting loops then I suppose it's an enhancement rather than a bug.

**Comment by jddxf** [comment]
It's unlikely to use something like `while (false)`, `if (true)` or `if (false)` while `while (true)` is a common construct. So I'm just fine with the current behavior.

**Comment by not-an-aardvark** [comment]
Closing this as it doesn't appear that there is enough interest, either from the team or from the community, to implement this at this time.
